docs/adr/adr-013-automatic-gas-price-estimation.md

```diff
@@ -0,0 +1,65 @@
+# ADR #013: Automatic Gas Price Estimation
+
+## Changelog
+
+- 17.01.2024: Initial design
+
+## Status
+
+Proposed
+
+## Context
+
+[CIP 18](https://github.com/celestiaorg/CIPs/blob/main/cips/cip-18.md) defines an interface that a light node can use for transaction submission to retrieve an estimate of the gas price and amount of gas required to include their transaction on Celestia. This predominantly is to allow for greater dynamism during times of congestions where gas prices become greater than the global minimum.
+
+This ADR defines how the go light node will interact with the service as a client. Note this does not cover how the server will actually estimate the gas prices and gas required.
+
+## Decision
+
+Allow users of Celestia nodes to opt in to automatic gas price and gas usage estimation for every write method
+
+## Detailed Design
+
+### When to estimate the gas price
+
+Users can opt in to automatic price estimation by using the existing `TxConfig.isGasPriceSet`. Previously, if this was false, the `DefaultGasPrice` was used, now if a user isn't manually setting the gas price, the node will query it from the `GasEstimator` service.
+
+### When to estimate the gas required
+
+Depending on the message types submitted, the node will decide whether to also query for the amount of gas that is required to submit the transaction. In the case of the following messages, the gas can be reliably estimated using a simple linear model:
+
+- `MsgPayForBlob`
+
+For all other messages (or combination of messages), the node will call the `GasEstimator` service.
+
+### Setting guardrails
+
+A new field in the `TxConfig` will be introduced allowing for a user to  use the gas price estimation but setting a maximum gas price as a cap that the user is willing to pay. If the price returned by the estimator is greater than the cap, likely due to a price spike, an error will be returned informing the user that the gas price required is currently greater than the user is willing to pay
+
+### Defaults
+
+By default, the same consensus node that the node is connected with, should be used as the `GasEstimator`. During construction, a user can optionally provide a grpc address that can be used instead.
+
+### Fallback
+
+If there are any errors with interacting with the `GasEstimator` service, the node should log an error and fallback to using the minimum gas price. 
+
+## Alternative Approaches
+
+There has been some discussion of a pricing mechanism built in to the protocol. This would be more complex but would remove the reliance on a trusted third party. This may be explored in the future but given resource constraints, the `GasEstimator` service has been prioritized
+
+## Consequences
+
+### Positive
+
+- Light nodes should 
+
+### Negative
+
+- Extra round trip with either the consensus node or someone offering the `GasEstimator` service before the transaction is submitted
```

Comment:
Would it make sense to subscribe to the `GasEstimator` and store the latest value?

Reply:
Hmm it's an option. But most light clients running will seldom be sending transactions, so imagine if a server had to constantly send updates of the gas price to these light clients which were only querying the state of their rollup/s and not actually submitting any transactions

Reply:
@cmwaters the core accessor (state service) in node is not only used by LNs but also FNs and BNs, so anyone using node to submit txs and has specified the default to use GasEstimator will have to go through this roundtrip. So maybe LNs will be seldom for submission, but FNs + BNs will be frequent.

Something to consider.

Reply:
If its seldom, this can be optimized and disable susbcribtion after certain period. 

Although, I am not sure that's worth the complexity of subscriptions for this case:
* This service is a short term solution
* We don't have evidence the additional rountrip is harming our users and we can always subscription once its  really needed. 

P.S. In other world where this Service is a long term solution, I would push for the subscription,)

Reply:
I get it's used by all node types. I would keep it simple for now and if it becomes problematic we can come up with a few optimizations or even change it to a subscription model.

Reply:
In this case, we should definitely have some spans inside submission process to measure total time as a metric so we can have some data from the get-go.